tools/shared.py

```diff
@@ -2994,13 +3006,22 @@ def add_emscripten_metadata(js_file, wasm_file):
       WebAssembly.lebify(EMSCRIPTEN_ABI_MAJOR) +
       WebAssembly.lebify(EMSCRIPTEN_ABI_MINOR) +
 
-      # static bump
       WebAssembly.lebify(mem_size) +
 
-      # table size
-      WebAssembly.lebify(table_size)
+      WebAssembly.lebify(table_size) +
+
+      WebAssembly.lebify(global_base) +
+
+      WebAssembly.lebify(dynamic_base) +
+
+      WebAssembly.lebify(dynamictop_ptr) +
+
+      WebAssembly.lebify(tempdouble_ptr) +
```

Comment:
Would it make more sense to export these things as globals from the wasm file?

If we are going to make this an ABI can we document to exact meaning on each of these things?

Do all these make sense with both fastcomp and the llvm backend?

Reply:
The reason they are in a different metadata section is that they must parsed to properly parameterize the runtime before the WASM can be successfully instantiated. (There is a circular dependence).

Medium term this will be documented exactly. I still don't consider it stable enough to document, which is partly the reason it isn't emitted automatically and requires "-s EMIT_EMSCRIPTEN_METADATA=1" to opt in. There is some initial documentation for developers here https://github.com/emscripten-core/emscripten/wiki/WebAssembly-Standalone#jswasm-abi

At least as far as emulating the JS runtime, `tempDoublePtr`, 'DYNAMICTOP_PTR`, and `DYNAMIC_BASE` makes sense with the LLVM backend. It wouldn't be possible for the host to implement `sbrk()` and `copyTempDouble()` without those parameters. `global_base` does not seem to be needed by LLVM generated WASM files (in asmjs/binaryen it's imported as `__memory_base`).

Reply:
I'm OK with landing this now, as long as we can iterate on this, change it, and hopefully simplify it in the future.    I wouldn't want to be locked in the ad-hoc conventions we currently have without a more formal spec.

Reply:
+1, Major version on this metadata section is still 0, so it's still explicitly not supported and not guaranteed to be in any specific format.